File: internal/agent/model_service.go
package agent

import (
	"context"
	"sync"

	aikit "github.com/Volpestyle/ai-kit/packages/go"
	"github.com/yourorg/coding-agent-harness/internal/config"
)

type ModelService struct {
	Kit          *aikit.Kit
	settingsPath string

	policyMu sync.RWMutex
	policy   config.ModelPolicy

	runner        *Runner
	sessionRunner *SessionRunner
}

func NewModelService(kit *aikit.Kit, policy config.ModelPolicy, settingsPath string, runner *Runner, sessionRunner *SessionRunner) *ModelService {
	return &ModelService{
		Kit:           kit,
		settingsPath:  settingsPath,
		policy:        policy,
		runner:        runner,
		sessionRunner: sessionRunner,
	}
}

func (s *ModelService) ListModels(ctx context.Context) ([]aikit.ModelRecord, error) {
	if s.Kit == nil {
		return nil, nil
	}
	return s.Kit.ListModelRecords(ctx, nil)
}

func (s *ModelService) GetPolicy() config.ModelPolicy {
	s.policyMu.RLock()
	defer s.policyMu.RUnlock()
	return s.policy
}

func (s *ModelService) SetPolicy(policy config.ModelPolicy) error {
	s.policyMu.Lock()
	s.policy = policy
	s.policyMu.Unlock()

	if s.runner != nil {
		s.runner.SetPolicy(policy)
	}
	if s.sessionRunner != nil {
		s.sessionRunner.SetPolicy(policy)
	}
	return config.SaveSettings(s.settingsPath, config.Settings{ModelPolicy: policy})
}

File: internal/agent/model_service.ts
import type { Kit, ModelRecord } from "@volpestyle/ai-kit-node";
import type { ModelPolicy } from "../config";
import { saveSettings } from "../config/settings";
import type { Runner } from "./runner";
import type { SessionRunner } from "./session_runner";

export class ModelService {
  private policy: ModelPolicy;

  constructor(
    private kit: Kit,
    policy: ModelPolicy,
    private settingsPath: string,
    private runner?: Runner,
    private sessionRunner?: SessionRunner,
  ) {
    this.policy = policy;
  }

  async listModels(): Promise<ModelRecord[]> {
    if (!this.kit) return [];
    return this.kit.listModelRecords();
  }

  getPolicy(): ModelPolicy {
    return this.policy;
  }

  async setPolicy(policy: ModelPolicy): Promise<void> {
    this.policy = policy;
    this.runner?.setPolicy(policy);
    this.sessionRunner?.setPolicy(policy);
    await saveSettings(this.settingsPath, { model_policy: policy });
  }
}
